packages/lib/services/joplinCloudUtils.ts

```diff
@@ -0,0 +1,104 @@
+import { Reducer } from 'react';
+import Setting from '../models/Setting';
+import { ApplicationPlatform, ApplicationType } from '../types';
+import shim from '../shim';
+import { _ } from '../locale';
+import { reg } from '../registry';
+
+type Events = 'LINK_USED' | 'COMPLETED';
+
+type IntitialValues = {
+	className: 'text' | 'bold';
+	message: string;
+	next: Events;
+	active: Events | 'INITIAL';
+};
+
+export const intitialValues: IntitialValues = {
+	className: 'text',
+	message: _('Waiting for authorisation...'),
+	next: 'LINK_USED',
+	active: 'INITIAL',
+};
+
+export const reducer: Reducer<IntitialValues, Events> = (state: IntitialValues, action: Events) => {
+	switch (action) {
+	case 'LINK_USED': {
+		return {
+			className: 'text',
+			message: _('If you have already authorised, please wait for the application to sync to Joplin Cloud.'),
+			next: 'COMPLETED',
+			active: 'LINK_USED',
+		};
+	}
+	case 'COMPLETED': {
+		return {
+			className: 'bold',
+			message: _('You are logged in into Joplin Cloud, you can leave this screen now.'),
+			active: 'COMPLETED',
+			next: 'COMPLETED',
+		};
+	}
+	default: {
+		return state;
+	}
+	}
+};
+
+export const getApplicationInformation = async () => {
+	const platformName = await shim.platformName();
+	switch (platformName) {
+	case 'ios':
+		return { type: ApplicationType.Mobile, platform: ApplicationPlatform.Ios };
+	case 'android':
+		return { type: ApplicationType.Mobile, platform: ApplicationPlatform.Android };
+	case 'darwin':
+		return { type: ApplicationType.Desktop, platform: ApplicationPlatform.MacOs };
+	case 'win32':
+		return { type: ApplicationType.Desktop, platform: ApplicationPlatform.Windows };
+	case 'linux':
+		return { type: ApplicationType.Desktop, platform: ApplicationPlatform.Linux };
+	default:
+		return { type: ApplicationType.Unknown, platform: ApplicationPlatform.Unknown };
+	}
+};
+
+export const generateLoginWithUniqueLoginCode = async (loginUrl: string, uniqueloginCode: string) => {
+	const applicationInfo = await getApplicationInformation();
+	const searchParams = new URLSearchParams();
+	searchParams.append('unique_login_code', uniqueloginCode);
+	searchParams.append('platform', applicationInfo.platform.toString());
+	searchParams.append('type', applicationInfo.type.toString());
+	searchParams.append('version', shim.appVersion());
+
+	return `${loginUrl}?${searchParams.toString()}`;
+};
+
+export const checkIfLoginWasSuccessful = async (applicationsUrl: string, ulc: string) => {
+	try {
+		const response = await fetch(`${applicationsUrl}?unique_login_code=${ulc}`);
+		if (!response) return undefined;
+
+		if (response.status === 200) {
+			const jsonResponse = await response?.json();
+			if (jsonResponse && (jsonResponse.id && jsonResponse.password)) {
```

Comment:
I feel the logic of this function is a bit off. Like why would the server respond successfully but not include the necessary information. Or if that info is missing, why wouldn't we want to notify the user instead of ignoring the response?

Then below you would call `response.json()` a second time which would throw an error (you can read the request body only once).

Also we shouldn't sprinkle `?` in our code "just in case" because it's going to silence type errors. And in fact you already have a check for the response above so it cannot be undefined at this point.

Have a look how other fetch calls are handled, I think the right flow is a check for `response.ok`, then get `response.json()` once and process this. With this there should be less conditionals and less defensive programming.

Reply:
You are right, I should have been more mindful when writing this code. I changed quite a bit, removed the try catch from this function and move to the caller (JoplinCloudLoginScree.tsx) so we can cancel the setInterval and show the error message to the user.